### PR TITLE
feat: [DHIS2-15463] Use dhis2 ui calendarInput component in working list

### PR DESCRIPTION
**Pull request by alaa-yahia**
Implements [DHIS2-15463](https://dhis2.atlassian.net/browse/DHIS2-15463)

[DHIS2-15463]: https://dhis2.atlassian.net/browse/DHIS2-15463?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by simonadomnisoru — COMMENTED** [review]
Hi @alaa-yahia,
I think there are 3 calendar-related Cypress tests in the working list that fail in [EventWorkingListsUser.feature](https://github.com/dhis2/capture-app/actions/runs/9899176366/job/27347760949?pr=3712). Can you take a look at them?
Support for editable input and dd-mm-yyyy format is needed, therefore this PR is blocked until https://github.com/dhis2/ui/pull/1504 is merged.

**Comment by alaa-yahia** [comment]
> Hi @alaa-yahia, I think there are 3 calendar-related Cypress tests in the working list that fail in [EventWorkingListsUser.feature](https://github.com/dhis2/capture-app/actions/runs/9899176366/job/27347760949?pr=3712). Can you take a look at them? Support for editable input and dd-mm-yyyy format is needed, therefore this PR is blocked until [dhis2/ui#1504](https://github.com/dhis2/ui/pull/1504) is merged.

Hey @simonadomnisoru
I have updated the tests.
An issue persists where the calendar popup remains open when switching focus between input fields, only closing when clicking elsewhere.
I will investigate it further and update the tests if I was able to solve it.

**Comment by simonadomnisoru** [comment]
> Hey @simonadomnisoru I have updated the tests. An issue persists where the calendar popup remains open when switching focus between input fields, only closing when clicking elsewhere. I will investigate it further and update the tests if I was able to solve it.

Hi Alaa, I noticed a few things that I believe need some improvement:
1) The calendar popup also stays open when you manually enter the date and then press Enter (perhaps it is the same issue you mentioned when switching focus between input fields).
2) While manually entering a date, the error "`Date string is invalid, received`" is displayed. Is it possible to only display this error after the user has finished typing (onBlur), the same way it currently works in play?
Screenshot 2024-08-07 at 12 42 43


3) When the calendar is in the `dd-mm-yyyy` format and you select a date, it is displayed in the input in the wrong format and the error "`Please provide a valid date`" is visible. When you manually type in a date in the `dd-mm-yyyy` format it works fine. In the video bellow you can see more details.
https://github.com/user-attachments/assets/6cfd254f-b161-4ae2-a1d4-660a26b


Let me know if you need any help. Thank you!

File: package.json
Comment (simonadomnisoru) [inline package.json:22]: Alphas releases are useful when developing, but IMHO it is best to wait for the stable `@dhis2/ui` release before merging this PR. Do you know the timeline/steps to get the stable release to include the alpha changes?

File: src/core_modules/capture-core/components/FormFields/DateAndTime/D2Date/D2Date.component.js
Comment (simonadomnisoru) [inline src/core_modules/capture-core/components/FormFields/DateAndTime/D2Date/D2Date.component.js:39]: There are use cases when the user wants to filter by only one date at a time. Therefore this `if` statement should only check `selectedDate !== undefined` to allow the empty string too.

Screenshot 2024-08-12 at 10 29 51

Thanks!

File: package.json
Comment (alaa-yahia) [reply package.json:22]: I was given the task to update the alpha with latest master. But I am not sure about the timeline as there are still accessibility task for CalendarInput.

File: src/core_modules/capture-core/components/FormFields/DateAndTime/D2Date/D2Date.component.js
Comment (alaa-yahia) [reply src/core_modules/capture-core/components/FormFields/DateAndTime/D2Date/D2Date.component.js:39]: I don't understand the requirements here ? Empty strings wouldn't be allowed as they aren't valid `selectedDate` and the selectedDate is `undefined` when nothing passed.

Comment (simonadomnisoru) [reply src/core_modules/capture-core/components/FormFields/DateAndTime/D2Date/D2Date.component.js:39]: The requirement is to allow empty strings because there are valid values. In the video below, there are steps for one of the use cases when the user clears the date input and then should be able to filter the working lists.

https://github.com/user-attachments/assets/177d662d-aad8-42e0-bb02-30e9cea79737



- `if (selectedDate)` - does not allow any [falsy](https://developer.mozilla.org/en-US/docs/Glossary/Falsy) value to pass
- `if (selectedDate !== undefined)` - does not allow `undefined` to pass

Let me know if it's still unclear and we can have a call to discuss further. Thank you 😄 !

Comment (alaa-yahia) [reply src/core_modules/capture-core/components/FormFields/DateAndTime/D2Date/D2Date.component.js]: Thanks for the video. I have update the code.

**Review by simonadomnisoru — APPROVED** [review]
LGTM! 👏

**Review by geethaalwan — APPROVED** [review]
Tested successfully on 2.42,2.41.2,2.40.5,2.39.7 versions

File: src/core_modules/capture-core/components/FiltersForTypes/Date/DateFilter.component.js
Comment (simonadomnisoru) [inline src/core_modules/capture-core/components/FiltersForTypes/Date/DateFilter.component.js:180]: The `The From date can't be after the To date` error is missing when the calendar is in the dd-MM-yyyy format. Below is a video with the steps on the branch vs play. 

https://github.com/user-attachments/assets/c914412c-e728-4c78-a563-1a3c67c827bb

Comment (JoakimSM) [reply src/core_modules/capture-core/components/FiltersForTypes/Date/DateFilter.component.js:180]: Probably ok to keep using moment for now? Why did we change this part?

**Review by JoakimSM — COMMENTED** [review]
Fix the bug @simonadomnisoru mentioned and I think this is ready. Good work!

File: src/core_modules/capture-core/components/FiltersForTypes/Date/DateFilter.component.js
Comment (alaa-yahia) [reply src/core_modules/capture-core/components/FiltersForTypes/Date/DateFilter.component.js:180]: @JoakimSM Because I wanted to remove `parseDate` function from parsers folder at the end if it isn't used anywhere else.

Comment (alaa-yahia) [reply src/core_modules/capture-core/components/FiltersForTypes/Date/DateFilter.component.js:180]: I put it back since it results in invalid dates for dd-mm-yyyy format

**Review by simonadomnisoru — APPROVED** [review]
It looks good to me now, well done @alaa-yahia! 👏

**Review by JoakimSM — APPROVED** [review]
LGTM 👍

**Pull request merged**
